Prod backend configurable

facundobatista opened the pull request:
This draft branch proposes two changes at once. After we settle some design decisions, I will cut both changes in different PRs, but for now I'm putting this here so it serves as example code.

Changes:

- Make the server backend urls configurable

- Have a very simple configuration layer to hold (still optional) charmcraft.yaml

facundobatista commented:
Related issue: #55

cjp256 commented on charmcraft/config.py:
```diff
@@ -0,0 +1,54 @@
+    """Hold the config from charmcraft.yaml."""
+
+    def __init__(self):
+        self.data = {}
```
I've been pushing Snapcraft away from (ab)using dictionaries and instead making this type of object into first-class python object to make it easier to use/maintain. Something like:

```
class Config:
  def __init__(self, *, name: str) -> None:
    self.name = name

  def validate(self) -> None:
    """Validate all properties."""
    ...

  @classmethod
  def unmarshal(self, raw_data) -> "Config":
    """Validate the schema & mypy typing assumptions and return Config."""
    if not isinstance(raw_data, dict):
            raise CommandError("Invalid charmcraft.yaml structure: must be a dictionary.")
    name = raw_data.get("name")
    if not isinstance(name, str):
            raise CommandError(f"Charm name {name!r} is not a valid string.")
    config = Config(name=name)
    config.validate()
    return config
````

facundobatista replied:
As we discussed, I'll go to a schema oriented complex object for this config.

cjp256 commented on charmcraft/config.py:
```diff
@@ -0,0 +1,54 @@
+        self.data = content
+
+
+config = _Config()
```
I would reconsider this global variable.  I expect it will cause unnecessary coupling.  Maybe instead just pass `config` as required, or limit to the properties relevant the user's need?

cmatsuoka replied:
I second that. This will make a difference if for some reason you need to refactor a subsystem into a reusable library in the future.

facundobatista replied:
As we discussed, I'll go to a model where just the minimum config is passed around.

cmatsuoka commented on charmcraft/utils.py:
```diff
@@ -33,6 +39,20 @@ def make_executable(fh):
+        with fpath.open('rb') as fh:
+            content = yaml.safe_load(fh)
+    except (yaml.error.YAMLError, OSError) as err:
+        logger.error("Failed to read/parse config file %s (got %r)", fpath, err)
```
Perhaps simply displaying the original error message after a colon would be easier to read ("Failed to do something: this bad thing happened").

facundobatista replied:
nice, improved

cmatsuoka commented on charmcraft/main.py:
```diff
@@ -192,32 +198,56 @@ def _pre_parse_args(self, sysargs):
+                    try:
+                        value = next(sysargs)
+                    except StopIteration:
+                        raise CommandError("The 'from' option expected one argument.")
```
Maybe "expects" instead of "expected"? Or "requires an argument"?

facundobatista replied:
much better, thanks!

facundobatista commented:
@cjp256 @cmatsuoka current code applies the two concepts we talked about: a hard-defined schema for the config, hold by a frozen complex object, and then passing around the minimum of what's needed, not having everything available everywhere.

I also transformed the just-landed Pack command for bundles to use this config (it previously had to deal directly with charmcraft.yaml).

Of course, this branch is not to be landed as is (that's why Ieft it in Draft), but the idea is to have approval of the general direction, and then I'll start cutting branches from here. First one will be just the config infrastructure with tests (without using it in that branch).

cjp256 commented on charmcraft/commands/pack.py:
```diff
@@ -114,18 +88,14 @@ def run(self, parsed_args):
             raise CommandError(
-                "Invalid charmcraft config; 'type' must be 'bundle' in file '{}'."
-                .format(charmcraft_filepath))
+                "Bad config; charmcraft.yaml missing or with an invalid 'type' field: "
```
Consider splitting up missing vs. unsupported type.

facundobatista replied:
Ok, will make this a more specific requirement.

cjp256 commented on charmcraft/config.py:
```diff
@@ -0,0 +1,152 @@
+    project_dirpath = attr.ib(default=None)
+
+    @classmethod
+    def from_file(cls, project_directory):
```
deconflict "from_file" vs directory argument. perhaps use the chamcraft file path as arg?

facundobatista replied:
Will decouple it even more. I will put a "load" function at module level.

cjp256 commented on charmcraft/config.py:
```diff
@@ -0,0 +1,152 @@
+        elif not isinstance(content, dict):
+            raise CommandError("Invalid charmcraft.yaml structure: must be a dictionary.")
+
+        content['project_dirpath'] = project_directory
```
consider decoupling this from the Config (maybe there's a case either for testing or future cases where you need to create configs without a project path readily available).  I'd perhaps consider a 'Project' class that encompasses these aspects of the project that aren't specifically detailed in the yaml.

sergiusens replied:
I am +1 on decoupling file finding (and opening) from config creation

facundobatista replied:
IIUC you're talking about different things here.

On one side, the config will hold a generic `project` option that would hold all project related details, like `directory`, and could be extended in the future.

On the other hand, I'll decouple the config loading from the config itself.

cjp256 commented on charmcraft/main.py:
```diff
@@ -95,11 +97,14 @@ def run(self, parsed_args, all_commands):
-    ('verbose', '-v', '--verbose', "Show debug information and be more verbose"),
-    ('quiet', '-q', '--quiet', "Only show warnings and errors, not progress"),
+# global options: the name used internally, its type, short and long parameters, and help text
+_Global = namedtuple('Global', 'name type short_option long_option help_message')
```
Not important, but dataclass (or attr) would be a better choice than namedtuple imo (type annotations, etc.)

facundobatista replied:
It's simpler like this and good enough for now, we can improve it in the future.

sergiusens commented on charmcraft/config.py:
```diff
@@ -0,0 +1,152 @@
+
+def _check_typefield(config, field, value):
+    """Check that the value of 'type' field is valid."""
+    # None is allowed until we make mandatory this configuration field
```
```suggestion
    # None is allowed until we make this configuration field mandatory
```

facundobatista replied:
thanks, improved.

sergiusens commented on charmcraft/config.py:
```diff
@@ -0,0 +1,152 @@
+    @classmethod
+    def from_dict(cls, parts):
+        """Build from a dicts sequence."""
+        if not isinstance(parts, dict):
```
one advantage of using jsonschema here is that you eventually get autocomplete on IDEs (VS Code, PyCharms, Emacs, etc)

facundobatista replied:
Will give it a try to `jsonschema`, I think all could be simpler and more declarative when using this.

sergiusens commented on charmcraft/config.py:
```diff
@@ -0,0 +1,152 @@
+                "Bad charmcraft.yaml content; the 'parts.bundle.prime' field must be a list: "
+                "got {!r}.".format(type(prime).__name__))
+
+        # validate that all are relative
```
Should this not handle:

```
$ ls prime
foo/ bar foobar
```

- Only Include foo and everything inside it
```yaml
prime:
    - foo
```

- Exclude foo and everything inside it (keeping bar and foobar)
```yaml
prime:
    - -foo
```

sergiusens replied:
If so, spec[0] would need some extra logic

facundobatista replied:
I don't like this "everything but this" rule, it suffers the same problem than "including everything by default": today you exclude `foo` and get `bar` and `foobar`, tomorrow you also add some confidential keys file and it will be included in the charm.

That said, note that this "basic prime" is a simple intermediate thing that will lead us towards the "full lifecycle capabilities", it doesn't need to cover all the cases.

facundobatista commented:
@cmatsuoka @sergiusens @cjp256 

As per the last round of reviews, I did the following changes:

- The Pack command now separates the cases of "missing a charmcraft.yaml config" (which is mandatory for this command) and the `type` field not being `bundle`. Note that it doesn't check that the field is there (this is handled in the config, as that field is mandatory).

- Removed `project_dirpath` in favor of a first class `project` entry that currently just holds `dirpath` but it may grow in the future.

- Refactored the whole config infrastructure to now use `jsonschema`; some details about this:

    - had to adapt the validation/format errors produced by that lib; surprisingly the adaptor function is small and without magic, and the error messages produced are the same than before using the library (this could grow/change in the future, though, if we need different messages -- probably not, but looking at how the jsonschema errors are massaged in snapcraft, we will see).

    - there's lot less manual checks, as `jsonschema` handled them; the only function in that regard that survived the refactor is the one validation that urls are as we need.

    - the config classes are smaller and neater; I even thought if we should keep these "attr classes" or we could go towards something simpler, and took the decission of leaving them as probably we will replace them by "dataclasses" when using a modern enough Python.

    - sure, now we have a splicit schema that we need to maintain, but this is **better**, as it's easier to understand than fields scattered all over the place; note that all the schema is unittested, specially to validate that the messages we're producing to the users are the one that we want.

    - decoupled the config creation from the config classes, now it's done through module function.

cjp256 commented on charmcraft/commands/store/client.py:
```diff
@@ -164,8 +159,14 @@ def _storage_push(monitor):
     """Lightweight layer above _AuthHolder to present a more network oriented interface."""
 
-    def __init__(self):
+    # default production URLs
```
Also, you may find it nicer to set the parameters with their respective default rather than write as a class/instance var.  Then when creating a client, the parameters are optional (nice for testing etc).

facundobatista replied:
As discussed, we'll defer using docstrings for sphinx-generated documentation for a future when charmcraft is a library (we're not generating docs from docstrings as it's only a CLI tool).

cjp256 commented on tests/conftest.py:
```diff
@@ -60,3 +62,27 @@ def __getattribute__(self, name):
+    class TestConfig(config_module._Config):
+        """The Config, but with a method to set test values."""
+
+        def set(self, **kwargs):
```
IMO, now that Config is a proper object, this fixture should no longer be used.  The tests should instantiate their own Config() with the necessary parameters required for that test rather than jump through this hoop.

facundobatista replied:
As the config is not omnipresent anymore, the config needs to be used in ~130 tests; having it as a fixture, easy to use in most of the cases where a special value is not needed reduces a lot the code needed in all the tests.

facundobatista commented on charmcraft/commands/store/client.py:
```diff
@@ -139,9 +134,9 @@ def request(self, method, url, body):
 
 
-def _storage_push(monitor):
+def _storage_push(monitor, storage_base_url):
```
I agree only when the multiple parameters are options.

facundobatista commented on charmcraft/config.py:
```diff
@@ -0,0 +1,170 @@
+}
+
+
+def adapt_validation_error(error):
```
We'll go "full type hinting" in the future. After that, all code landed should have type hints, yes.

sergiusens commented on charmcraft/config.py:
```diff
@@ -0,0 +1,170 @@
+            },
+        },
+    },
+    'required': ['type'],
```
this will help with the top level entries, we should not do this for parts though

facundobatista replied:
Good idea. Will do.

sergiusens reviewed (commented):
this should move forward into its smaller PRs, general structure looks good

facundobatista commented on charmcraft/config.py:
```diff
@@ -0,0 +1,170 @@
+    # XXX Facundo 2021-01-04: we will make this configuration mandatory in the future, but
+    # so far is ok to not have it
+    if content is None:
+        return
```
I don't like explicit None in empty returns :/

facundobatista commented:
All this went in (except #229, but already proposed).
